include/wlr/types/wlr_text_input.h

```diff
@@ -0,0 +1,57 @@
+#ifndef WLR_TYPES_WLR_TEXT_INPUT_H
+#define WLR_TYPES_WLR_TEXT_INPUT_H
+
+#include <wayland-server.h>
+#include <wlr/types/wlr_seat.h>
+#include <wlr/types/wlr_surface.h>
+
+struct wlr_text_input_atomic_properties {
+	struct {
+		char *text;
+		uint32_t cursor;
+		uint32_t anchor;
+	} surrounding;
+	struct {
+		uint32_t hint;
+		uint32_t purpose;
+	} content_type;
+	struct {
+		int32_t x;
+		int32_t y;
+		int32_t width;
+		int32_t height;
+	} cursor_rectangle;
+};
+
+struct wlr_text_input {
+	struct wlr_seat *seat; // used only for unlinking seat->text_input; perhaps better to use the destroy signal
```

Comment:
This sort of comment is better put into the commit message or here on GitHub

Reply:
Sorry, forgot to move it before committing. What do you think about the idea in the comment anyway?

Reply:
Probably better to use the destroy signal, yeah.

Reply:
I don't think so. That would require adding text_input code in the seat.

Reply:
Ah, good point. This patch already touches the seat, should probably be cleaned out.

Reply:
Seat code is touched because there needs to be some way to get notified of new `text_input` instances. Since only one per seat is allowed, I thought seat is [the best place](https://github.com/swaywm/wlroots/pull/776/files#diff-6eb821df1a7fff48e265e52e0a742741R215). Should it be placed somewhere else instead?

Reply:
Probably better to construct a text_input with a seat reference and have it manage it. Or let the compositor glue stuff together.